interp: cannot type assert from empty interface into slice

The following program `sample.go` triggers a panic:

```go
package main

import (
	"fmt"
)

func interfaceAsInts() {
	var a interface{}
	b := 2
	c := 3
	a = []int{b, c}

	d, ok := a.([]int)
	if !ok {
		println("nope")
		return
	}

	for _, v := range d {
		fmt.Println(v)
	}
}

func interfaceAsInterfaces() {
	var a, b, c interface{}
	b = 2
	c = 3
	a = []interface{}{b, c}

	d, ok := a.([]interface{})
	if !ok {
		println("nope")
		return
	}

	for _, v := range d {
		fmt.Println(v)
	}
}

func main() {
	interfaceAsInts()
	interfaceAsInterfaces()
}
```

Expected result:
```console
$ go run ./sample.go
2
3
2
3
```

Got:
```console
$ yaegi ./sample.go
_test/composite15.go:11:6: panic
_test/composite15.go:43:2: panic
run: reflect.Set: value of type []int is not assignable to type interp.valueInterface
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc0002e0800, 0x40)
	/Users/mpl/go1/src/runtime/debug/stack.go:24 +0x9f
github.com/traefik/yaegi/interp.(*Interpreter).eval.func1(0xc0001bbc70)
	/Users/mpl/src/github.com/traefik/yaegi/interp/interp.go:480 +0xc5
panic(0x19c11c0, 0xc000207760)
	/Users/mpl/go1/src/runtime/panic.go:969 +0x1b9
github.com/traefik/yaegi/interp.runCfg.func1(0xc0003640a0, 0xc000471a00, 0xc0001bba60)
	/Users/mpl/src/github.com/traefik/yaegi/interp/run.go:184 +0x247
panic(0x19c11c0, 0xc000207760)
	/Users/mpl/go1/src/runtime/panic.go:969 +0x1b9
github.com/traefik/yaegi/interp.runCfg.func1(0xc000364140, 0xc000468d00, 0xc0001bb840)
	/Users/mpl/src/github.com/traefik/yaegi/interp/run.go:184 +0x247
panic(0x19c11c0, 0xc000207760)
	/Users/mpl/go1/src/runtime/panic.go:969 +0x1b9
reflect.Value.assignTo(0x19b5880, 0xc000205ca0, 0x97, 0x1afb136, 0xb, 0x1a43620, 0x0, 0x19c0480, 0xc00020be88, 0x182)
	/Users/mpl/go1/src/reflect/value.go:2425 +0x405
reflect.Value.Set(0x1a43620, 0xc000205c80, 0x199, 0x19b5880, 0xc000205ca0, 0x97)
	/Users/mpl/go1/src/reflect/value.go:1554 +0xbd
github.com/traefik/yaegi/interp.arrayLit.func1(0xc000364140, 0xc0003619d0)
	/Users/mpl/src/github.com/traefik/yaegi/interp/run.go:2149 +0x225
github.com/traefik/yaegi/interp.runCfg(0xc000468d00, 0xc000364140)
	/Users/mpl/src/github.com/traefik/yaegi/interp/run.go:190 +0x87
github.com/traefik/yaegi/interp.call.func7(0xc0003640a0, 0xc000468000)
	/Users/mpl/src/github.com/traefik/yaegi/interp/run.go:1209 +0xce8
github.com/traefik/yaegi/interp.runCfg(0xc000471a00, 0xc0003640a0)
	/Users/mpl/src/github.com/traefik/yaegi/interp/run.go:190 +0x87
github.com/traefik/yaegi/interp.(*Interpreter).run(0xc00035e000, 0xc000471300, 0xc000364000)
	/Users/mpl/src/github.com/traefik/yaegi/interp/run.go:121 +0x2af
github.com/traefik/yaegi/interp.(*Interpreter).eval(0xc00035e000, 0xc0002e3680, 0x209, 0x7ffeefbff6d2, 0x14, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/mpl/src/github.com/traefik/yaegi/interp/interp.go:572 +0x809
github.com/traefik/yaegi/interp.(*Interpreter).EvalPath(0xc00035e000, 0x7ffeefbff6d2, 0x14, 0xc0002e3400, 0x209, 0x0, 0x0, 0xc000212ea0)
	/Users/mpl/src/github.com/traefik/yaegi/interp/interp.go:394 +0xfa
main.runFile(0xc00035e000, 0x7ffeefbff6d2, 0x14, 0xc000200360, 0x1)
	/Users/mpl/src/github.com/traefik/yaegi/cmd/yaegi/run.go:123 +0xae
main.run(0xc0000321d0, 0x1, 0x1, 0x13, 0x13)
	/Users/mpl/src/github.com/traefik/yaegi/cmd/yaegi/run.go:89 +0x7e5
main.main()
	/Users/mpl/src/github.com/traefik/yaegi/cmd/yaegi/yaegi.go:133 +0x479
```

--- a/_test/composite15.go
+++ b/_test/composite15.go
@@ -1,0 +1,50 @@
+package main
+
+import (
+	"fmt"
+)
+
+func interfaceAsInts() {
+	var a interface{}
+	b := 2
+	c := 3
+	a = []int{b, c}
+
+	d, ok := a.([]int)
+	if !ok {
+		println("nope")
+		return
+	}
+
+	for _, v := range d {
+		fmt.Println(v)
+	}
+}
+
+func interfaceAsInterfaces() {
+	var a, b, c interface{}
+	b = 2
+	c = 3
+	a = []interface{}{b, c}
+
+	d, ok := a.([]interface{})
+	if !ok {
+		println("nope")
+		return
+	}
+
+	for _, v := range d {
+		fmt.Println(v)
+	}
+}
+
+func main() {
+	interfaceAsInts()
+	interfaceAsInterfaces()
+}
+
+// Output:
+// 2
+// 3
+// 2
+// 3

--- a/interp/run.go
+++ b/interp/run.go
@@ -415,7 +415,14 @@ func typeAssert(n *node, withResult, withOk bool) {
 				}
 				return next
 			}
-			ok = canAssertTypes(v.value.Type(), rtype)
+
+			styp := v.value.Type()
+			// TODO(mpl): probably also maps and others. and might have to recurse too.
+			if styp.String() == "[]interp.valueInterface" {
+				styp = v.node.typ.rtype
+			}
+
+			ok = canAssertTypes(styp, rtype)
 			if !ok {
 				if !withOk {
 					panic(fmt.Sprintf("interface conversion: interface {} is %s, not %s", v.value.Type().String(), rtype.String()))
@@ -2153,7 +2160,11 @@ func arrayLit(n *node) {
 		for i, v := range values {
 			a.Index(index[i]).Set(v(f))
 		}
-		value(f).Set(a)
+		dest := value(f)
+		if _, ok := dest.Interface().(valueInterface); ok {
+			a = reflect.ValueOf(valueInterface{n, a})
+		}
+		dest.Set(a)
 		return next
 	}
 }

